great_expectations/checkpoint/actions.py

```diff
@@ -103,8 +108,79 @@ def filter_results(self, class_: Type[ValidationAction]) -> list[dict]:
         return [action_result for action, action_result in self._data if isinstance(action, class_)]
 
 
+class ValidationActionRegistry:
+    """
+    Registers both GX-defined and user-defined ValidationActions.
+
+    Enables the instantiation of ValidationActions based on their configuration.
+    """
+
+    def __init__(self):
+        self._registered_actions: dict[str, Type[ValidationAction]] = {}
+
+    def register(self, action_type: str, action_class: Type[ValidationAction]) -> None:
+        """
+        Register a ValidationAction class with the registry.
+
+        Args:
+            action_type: The type of the action to register.
+            action_class: The ValidationAction class to register.
+
+        Raises:
+            ValidationActionAlreadyRegisteredError: If the action type is already registered.
+        """
+        if action_type in self._registered_actions:
+            raise ValidationActionAlreadyRegisteredError(action_type)
+
+        self._registered_actions[action_type] = action_class
+
+    def get_from_config(self, action_config: dict) -> Type[ValidationAction]:
+        """
+        Return a ValidationAction class based on its configuration.
+        Used when instantiating actions from a checkpoint configuration.
+
+        Args:
+            action_config: The configuration of the action to instantiate.
+
+        Returns:
+            The ValidationAction class corresponding to the configuration.
+
+        Raises:
+            ValidationActionRegistryRetrievalError: If the action type is not registered.
+        """
+        action_type = action_config.get("type")
+        if action_type not in self._registered_actions:
+            raise ValidationActionRegistryRetrievalError(action_type)
+
+        return self._registered_actions[action_type]
+
+
+_VALIDATION_ACTION_REGISTRY = ValidationActionRegistry()
+
+
+@dataclass_transform(kw_only_default=True, field_specifiers=(Field,))
+class MetaValidationAction(ModelMetaclass):
+    """MetaValidationAction registers ValidationAction as they are defined, adding them to
+    the registry.
+
+    Any class inheriting from ValidationAction will be registered based on the value of the
+    "type" class attribute.
+    """
+
+    def __new__(cls, clsname, bases, attrs):
+        newclass = super().__new__(cls, clsname, bases, attrs)
+
+        action_type = newclass.__fields__.get("type")
+        if action_type and action_type.default:  # Excludes base classes
+            _VALIDATION_ACTION_REGISTRY.register(
+                action_type=action_type.default, action_class=newclass
+            )
+
+        return newclass
```

Comment:
Can this be done cleaner? I want to avoid `ValidationAction` and `DocsAction` base classes that don't have `type`